sklearn/utils/stats.py

```diff
@@ -25,37 +26,98 @@ def _weighted_percentile(array, sample_weight, percentile=50):
         sample_weight = np.tile(sample_weight, (array.shape[1], 1)).T
+
+    n_rows, n_cols = array.shape
+
     sorted_idx = np.argsort(array, axis=0)
     sorted_weights = _take_along_axis(sample_weight, sorted_idx, axis=0)
-
-    # Find index of median prediction for each sample
+    percentile = [percentile / 100] * n_cols
     weight_cdf = stable_cumsum(sorted_weights, axis=0)
-    adjusted_percentile = percentile / 100 * weight_cdf[-1]
-    percentile_idx = np.array([
-        np.searchsorted(weight_cdf[:, i], adjusted_percentile[i])
-        for i in range(weight_cdf.shape[1])
-    ])
-    percentile_idx = np.array(percentile_idx)
-    # In rare cases, percentile_idx equals to sorted_idx.shape[0]
-    max_idx = sorted_idx.shape[0] - 1
-    percentile_idx = np.apply_along_axis(lambda x: np.clip(x, 0, max_idx),
-                                         axis=0, arr=percentile_idx)
-
-    col_index = np.arange(array.shape[1])
-    percentile_in_sorted = sorted_idx[percentile_idx, col_index]
-    percentile = array[percentile_in_sorted, col_index]
-    return percentile[0] if n_dim == 1 else percentile
+
+    def _squeeze_arr(arr, n_dim):
+        return arr[0] if n_dim == 1 else arr
+
+    if interpolation == "nearest":
+        # compute by nearest-rank method
+        adjusted_percentile = percentile * weight_cdf[-1]
+        percentile_value_idx = np.array([
+            np.searchsorted(weight_cdf[:, i], adjusted_percentile[i])
+            for i in range(weight_cdf.shape[1])
+        ])
+
+        percentile_value_idx = np.apply_along_axis(
+            lambda x: np.clip(x, 0, n_rows - 1), axis=0,
+            arr=percentile_value_idx
+        )
+        percentile_value = array[
+            sorted_idx[percentile_value_idx, np.arange(n_cols)],
+            np.arange(n_cols)
+        ]
+        return _squeeze_arr(percentile_value, n_dim)
+
+    elif interpolation in ("linear", "lower", "higher"):
+        # compute by linear interpolation between closest ranks method
+        adjusted_percentile = (weight_cdf - 0.5 * sorted_weights)
+        with np.errstate(invalid="ignore"):
+            adjusted_percentile /= weight_cdf[-1]
+
+        if interpolation in ("lower", "higher"):
+            percentile_idx = np.array([
+                np.searchsorted(adjusted_percentile[:, col], percentile[col],
+                                side="left")
+                for col in range(adjusted_percentile.shape[1])
+            ])
+            if interpolation == "lower":
+                percentile_idx -= 1
+            percentile_idx = np.apply_along_axis(
+                lambda x: np.clip(x, 0, n_rows - 1), axis=0,
+                arr=percentile_idx
+            )
+            percentile = np.nan_to_num([
+                adjusted_percentile[percentile_idx[i], i]
+                for i in range(adjusted_percentile.shape[1])
+            ], nan=percentile)
+
+        percentile_value = np.array([
+            np.interp(
```

Comment:
`np.interp` seems not to implement the linear interpolation of `np.percentile`. Not sure what we should do then.